src/cascadia/TerminalControl/TermControlAutomationPeer.cpp

```diff
@@ -142,8 +177,62 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         });
     }
 
+    void TermControlAutomationPeer::NotifyNewOutput(std::wstring newOutput)
+    {
+        // Try to suppress any events (or event data)
+        // that is just the keypress the user made
+        auto sanitized{ Sanitize(newOutput) };
+        while (!_keyEvents.empty() && IsReadable(sanitized))
+        {
+            if (towupper(sanitized.front()) == _keyEvents.back())

```

Comment:
Do you want to uppercase unicode here? In that case `towupper` won't be accurate, because some languages need more or less code points when transforming to upper case (e.g. ß to SS).
If you only want to convert ASCII you can use `til::toupper_ascii`.

To be honest I don't quite get what `_keyEvents` does here. If I'm the only one, don't mind me, but if others feel the same way, a more comprehensive comment would probably be helpful.

Reply:
Added the following comment in:
```
                // the key event's character (i.e. the "A" key) matches
                // the output character (i.e. "a" or "A" text).
                // We can assume that the output character resulted from
                // the pressed key, so we can ignore it.
```

Reply:
Isn't that basically a race condition since the communication with the shell is asynchronous?

Reply:
But the key down event gets stored until new output comes out. So doesn't that force it to be synchronized?

Reply:
> But what if it gets out of sync? For instance what if the shell prints "hello" and I press "e" simultaneously so that it says "helelo". Wouldn't that result in "hlelo" instead of "hello"?

Hmm... I'm not sure if that would happen. But also, keyboard events interrupt the new output from being read, because that's what the screen reader just outright does. So, if that happens, I think it would immediately get interrupted by you pressing the 'e' key and the screen reader would just read 'e'.

(let me know if I'm not clear, it's a bit weird to explain haha)

Reply:
> keyboard events interrupt the new output from being read

By default, and in 99% of cases, that's true, but this can technically be disabled: disable "speak typed characters", "speak typed words", and "speech interrupt for typed characters" in NVDA's keyboard settings (<kbd>Ctrl+nvda+k</kbd>).